rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,350 @@
+- collection_id - identifies the collection
+- data set_id - identifies the data set
+
+##### Response: POST {collection_id}/{dataset_id}/upload/file
+
+- Access Token - allows the user to upload their file to S3 in multiple parts.
+- Presigned URL - allows the user to upload their file to S3 in a single part.
+- asset_uuid - identifies the name of the new file created.
+
+#### POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload
+
+Used to request an access token to continue uploading a file to S3 by an authenticated user who owns the data set. An
+error is returned if the file upload has been completed already.
+
+##### Request: POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload
+
+- data set_id - identifies the data set.
+- asset_id - identifies the asset.
+
+##### Response: POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload
+
+- Access Token - allows the user to upload their file to S3 in multiple parts.
+
+### AWS STS
+
+The AWS STS service will be used to generate temporary credential the user can use to upload a file to S3. The backend
+server will use the STS [assume role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html)
+functionality to generate a token with temporary access to upload to S3. According to AWS:
+
+> The resulting session's permissions are the intersection of the role's identity-based policies and the session policies.
+
+The IAM policies bellow will be combined to limit the access of the token generate to only allow the bearer to upload the
+file in the session policy resource. The session policy will be generated at the time of the request, and will use the
+asset_uuid generated by the server and the data set_uuid provided in the request.
+
+#### Assumed Role IAM Policy
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:ListMultipartUploadParts",
+        "s3:PutObject",
+        "s3:AbortMultipartUpload"
+      ],
+      "Resource": [
+        "arn:aws:s3:::upload-bucket/"
+      ]
+    }
+    ]
+}
+```
+
+#### IAM Session Policy
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:ListMultipartUploadParts",
+        "s3:PutObject",
+        "s3:AbortMultipartUpload"
+      ],
+      "Resource": [
+        "arn:aws:s3:::upload-bucket/{data set_uuid}/{asset_uuid}"
+      ]
+    }
+    ]
+}
+```
+
+### EC2
+
+The EC2 instance will be use to download files from a CSP and upload to our Upload bucket.
+### Upload Bucket
+
+The upload bucket is a separate bucket or separate bucket location within our main bucket. It is used to stage files
+before moving to the main bucket.
+
+S3 event can be use to trigger additional processing such as validation, once an upload has completed.
+
+### Cloud to Cloud Upload Flow
+
+Cloud to cloud uploads will allow a user to provide a sharable link from their CSP to the file they wish to upload,
```

Comment:
Perhaps a small rewrite - _Cloud to cloud uploads require the user to provide a shareable link for a file on their CSP from which DP can upload the file to the S3 bucket_